c/secp256k1_blake160_sighash_all.c

```diff
@@ -28,164 +25,128 @@
 
 /*
- * Arguments are listed in the following order:
- * 0 ~ n. pubkey blake160 hash, blake2b hash of pubkey first 20 bytes, used to
- * shield the real pubkey in lock script, the lock is considered as a multisig
- * lock when the number of pubkeys is more than 1.
- * n + 1. multisig threshold (optional), used to indicate the threshold of a
- * multisig lock, must ignore this arg when only one pubkey is used.
+ * Arguments:
+ * pubkey blake160 hash, blake2b hash of pubkey first 20 bytes, used to
+ * shield the real pubkey.
  *
  * Witness:
- * 0 ~ m signature, signatures used to present ownership, the number of signatures
- * must be equals to the threshold if the lock is a multisig lock, otherwise must
- * use only 1 signature.
+ * signature used to present ownership.
  */
 int main() {
   int ret;
-  size_t index = 0;
-  uint64_t sigs_cnt = 0;
-  uint64_t threshold = 0;
   volatile uint64_t len = 0;
-  unsigned char tx_hash[BLAKE2B_BLOCK_SIZE];
   unsigned char temp[TEMP_SIZE];
-  unsigned char witness[WITNESS_SIZE];
-  unsigned char script[SCRIPT_SIZE];
-  uint8_t secp_data[CKB_SECP256K1_DATA_SIZE];
-  mol_pos_t script_pos;
-  mol_read_res_t args_res;
-  mol_read_res_t bytes_res;
 
   /* Load args */
+  unsigned char script[SCRIPT_SIZE];
   len = SCRIPT_SIZE;
   ret = ckb_load_script(script, &len, 0);
   if (ret != CKB_SUCCESS) {
     return ERROR_SYSCALL;
   }
+  if (len > SCRIPT_SIZE) {
+    return ERROR_SCRIPT_TOO_LONG;
+  }
+  mol_pos_t script_pos;
   script_pos.ptr = (const uint8_t*)script;
   script_pos.size = len;
-  args_res = mol_cut(&script_pos, MOL_Script_args());
+
+  mol_read_res_t args_res = mol_cut(&script_pos, MOL_Script_args());
   if (args_res.code != 0) {
     return ERROR_ENCODING;
   }
-  bytes_res = mol_cut_bytes(&args_res.pos);
-  if (bytes_res.code != 0) {
+  mol_read_res_t args_bytes_res = mol_cut_bytes(&args_res.pos);
+  if (args_bytes_res.code != 0) {
     return ERROR_ENCODING;
-  } else if (bytes_res.pos.size == 0) {
-    return ERROR_WRONG_NUMBER_OF_ARGUMENTS;
+  } else if (args_bytes_res.pos.size != BLAKE160_SIZE) {
+    return ERROR_ARGUMENTS_LEN;
   }
 
-  /* Calculate sigs count and threshold */
-  sigs_cnt = bytes_res.pos.size / BLAKE160_SIZE;
-  if (bytes_res.pos.size % BLAKE160_SIZE == 0) {
-    threshold = sigs_cnt;
-  } else {
-    threshold = bytes_res.pos.ptr[sigs_cnt * BLAKE160_SIZE];
+  /* Load witness of first input */
+  unsigned char witness[WITNESS_SIZE];
+  len = WITNESS_SIZE;
+  ret = ckb_load_witness(witness, &len, 0, 0, CKB_SOURCE_GROUP_INPUT);
+  if (ret != CKB_SUCCESS) {
+    return ERROR_SYSCALL;
   }
-
-  if (threshold > sigs_cnt || threshold == 0) {
-    return ERROR_INVALID_THRESHOLD;
+  if (len > WITNESS_SIZE) {
+    return ERROR_WITNESS_TOO_LONG;
   }
-
-  uint8_t used_signatures[sigs_cnt];
-
-  secp256k1_context context;
-  ret = ckb_secp256k1_custom_verify_only_initialize(&context, secp_data);
-  if (ret != 0) {
-    return ret;
+  if (len < SIGNATURE_SIZE) {
```

Comment:
I don't know if I like the implicit rule here.

1. This lock script has a hidden rule that the first witness item has 65-byte long signature, then the rest data are for type scripts.
2. In NervosDAO script, we are using the last 8 bytes of the witness, which should be byte 65 - 72 here.
3. While theoretically, those 2 rules are decoupled, it's very easy to get confused, people might just assume that each lock script uses 65 bytes of the witness part, and the rest part is for type script. When this happens, we will have an unnecessary bundling between lock script and type script.

So if I'm working on this feature, I will introduce a length field(most likely 2 bytes) denoting the length of witness used by lock script, the remaining part will be used by the type script(if any). IMHO this is far more clear than a hidden 65-byte rule, which might change if lock uses a different signature algorithm, while a 2-byte length field works for more cases.

Reply:
Continuing on this problem, if my type script uses variable length witness data, which is not fixed 8 bytes witness like NervosDAO does, I will have problem integrating with the lock script here. I will have to do one of the following solutions:

1. Assume all lock scripts uses 65 fixed length signature;
2. Detect if the cell is using this particular lock, if so uses 65 bytes, otherwise use a different solution;

Both of them are quite dirty to me, and I believe the problem here, is that assuming 65 byte lock witness here, really does limit us to the way this lock script can corporate with other type scripts.

Reply:
> Continuing on this problem, if my type script uses variable length witness data, which is not fixed 8 bytes witness like NervosDAO does, I will have problem integrating with the lock script here. I will have to do one of the following solutions:

You can append type script's witness variable len to the end of witness, and read the last len bytes first in your type script contract.

Reply:
It does solve the variable length witness problem(tho arguably in a quirky way), but the first problem stands: when people see the first 65 bytes is a signature, they might just assume all witness has 65-byte signature at front, which might cause more problems. Personally I'd suggest build an alternative format to avoid the confusion, or maybe just introduce molecule to proper serialize witness instead of those minimal serialisation part which only works for several contracts